code/modules/projectiles/sentries.dm

```diff
@@ -0,0 +1,421 @@
+	if(!scan())
+		var/obj/item/weapon/gun/gun = internal_item
+		gun.stop_fire()
+		return
+	playsound(loc, 'sound/items/detector.ogg', 25, FALSE)
+
+	sentry_start_fire()
+
+///Checks the nearby mobs for eligability. If they can be targets it stores them in potential_targets. Returns TRUE if there are targets, FALSE if not.
+/obj/machinery/deployable/mounted/sentry/proc/scan()
+	var/obj/item/weapon/gun/gun = internal_item
+	potential_targets.Cut()
+	for (var/mob/living/carbon/human/nearby_human AS in cheap_get_humans_near(src, range))
+		if(nearby_human.stat == DEAD || CHECK_BITFIELD(nearby_human.status_flags, INCORPOREAL)  || (CHECK_BITFIELD(gun.turret_flags, TURRET_SAFETY) || nearby_human.wear_id?.iff_signal & iff_signal))
+			continue
+		potential_targets += nearby_human
+	for (var/mob/living/carbon/xenomorph/nearby_xeno AS in cheap_get_xenos_near(src, range))
+		if(nearby_xeno.stat == DEAD || CHECK_BITFIELD(nearby_xeno.status_flags, INCORPOREAL)) //So wraiths wont be shot at when in phase shift
+			continue
+		potential_targets += nearby_xeno
+	return potential_targets.len
+
+///Checks the range and the path of the target currently being shot at to see if it is eligable for being shot at again. If not it will stop the firing.
+/obj/machinery/deployable/mounted/sentry/proc/check_next_shot(datum/source, atom/gun_target, obj/item/weapon/gun/gun)
+	SIGNAL_HANDLER
+	var/obj/item/weapon/gun/internal_gun = internal_item
+	if(get_dist(src, gun_target) <= range && (CHECK_BITFIELD(get_dir(src, gun_target), dir) || CHECK_BITFIELD(internal_gun.turret_flags, TURRET_RADIAL)) && check_target_path(gun_target))
+		return
+	internal_gun.stop_fire()
+
+///Sees if theres a target to shoot, then handles firing.
+/obj/machinery/deployable/mounted/sentry/proc/sentry_start_fire()
+	var/obj/item/weapon/gun/gun = internal_item
+	var/mob/living/target = get_target()
+	update_icon_state()
+	if(target != gun.target || get_dist(src, target) > range)
+		gun.stop_fire()
+	if(!gun.current_mag)
+		sentry_alert(SENTRY_ALERT_AMMO)
+		return
+	if(CHECK_BITFIELD(gun.turret_flags, TURRET_RADIAL))
+		var/new_dir = get_dir(src, target)
+		switch(new_dir)
+			if(NORTHWEST)
+				new_dir = NORTH
+			if(NORTHEAST)
+				new_dir = EAST
+			if(SOUTHWEST)
+				new_dir = WEST
+			if(SOUTHEAST)
+				new_dir = SOUTH
+		setDir(new_dir)
+		if(gun.sentry_battery.charge <= 0)
+			sentry_alert(SENTRY_ALERT_BATTERY)
+	if(CHECK_BITFIELD(gun.flags_gun_features, GUN_BURST_FIRING))
+		gun.set_target(target)
+		return
+	gun.start_fire(src, target, bypass_checks = TRUE)
+
+///Checks the path to the target for obstructions. Returns TRUE if the path is clear, FALSE if not.
+/obj/machinery/deployable/mounted/sentry/proc/check_target_path(mob/living/target)
+	var/list/turf/path = getline(src, target)
+	path -= get_turf(src)
+	if(!path.len)
+		return FALSE
+	for(var/turf/T AS in path)
+		var/obj/effect/particle_effect/smoke/smoke = locate() in T
+		if(smoke)
+			return FALSE
```

Comment:
why any smoke instead of only vision blocking ones like it was before ? not every smoke blocks vision.